### PR TITLE
Proof-of-concept prototype of provenance (DM-3958)

### PR DESCRIPTION
**Pull request by jbecla**

### CODE REVIEW

File: calibPipe.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+
+import MySQLdb
```
Comment (andy-slac) [inline calibPipe.py:2]: Why not `lsst.db`? I thought it should be easier to use and configure than mysqldb?

File: calibPipe.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            cursor.execute('SELECT rawExposureId FROM RawExposure')
+            rows = cursor.fetchall()
+            for row in rows:
+                cursor.execute('''
+INSERT INTO ScienceCalibratedExposure(rawExposureId, filter, ra, decl,
+                                      obsStart, cFlux, cFluxErr)
+  SELECT r.rawExposureId, r.filter, r.ra, r.decl, r.obsStart, r.flux*c.v, %f
+  FROM   RawExposure r, RawCalibExposure c
+  WHERE  r.rawExposureId=%s AND r.filter=c.filter
+     AND r.ra=c.ra AND r.decl=c.decl''' % (fluxErr, row[0]))
```
Comment (andy-slac) [inline calibPipe.py:63]: This sort of indentation inside string makes it very hard to see how Python code is indented. Maybe better to move multi-line string literals outside method definition?

Comment (jbecla) [inline calibPipe.py:63]: yes, I agree and I kept playing with some alternatives. Would you make it a private class variable then? self._queryFindVersionOfTaskCalibA or something like that?

File: calibPipe.py (hunk 1 above)
Comment (jbecla) [inline calibPipe.py:2]: Note this is a very standalone prototype with no dependency on stack, it is not even in github.com:lsst, but in github.com:lsst-dm. I will switch once we make this part of stack. Deal?

File: provDetective.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        self._cursor.execute('''
+SELECT scExposureId, sourceId, sceGroupId, taskExecId, taskName, nodeName
+FROM   Source
+JOIN   ScienceCalibratedExposure USING(scExposureId)
+JOIN   prv_SCEExposureToGroup USING(scExposureId)
+JOIN   prv_TaskExecutionToSCEGroup USING(sceGroupId)
+JOIN   prv_TaskExecution USING(taskExecId)
+JOIN   prv_Node USING(nodeId)
+JOIN   prv_Task USING(taskId)
+WHERE  objectId=%s''' % objectId)
```
Comment (andy-slac) [inline provDetective.py:44]: This looks like it mixes qserv tables with provenance tables, does it mean that provenance schema has to be in qserv?

Comment (jbecla) [inline provDetective.py:44]: I thought it'd be available through qserv like any other non-partitioned tables. But I can guess why you are asking... provenance related queries are likely going to stretch out parser. I think we can work around it... expose provenance through non-qserv standalone mysql maybe? After all, the linkage between qserv-manage tables and provenance is pretty weak. Let's talk about it at our next meeting.

File: provProto.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  AND  t.taskName = "%s"
+  AND  validityEnd="%s"''' % (task.name, self._infinity))
+        taskId = cursor.fetchone()[0]
+        cursor.execute('''
+UPDATE prv_cnf_Task SET validityEnd="%s" WHERE taskId="%s"''' % \
+                           (self._currentTime, taskId))
+        cursor.execute('''
+INSERT INTO prv_cnf_Task(taskId, validityBegin, validityEnd, gitSHA)
+VALUES (%d, "%s", "%s", "%s")''' % \
+                (taskId, self._currentTime, self._infinity, task.gitSHA))
```
Comment (andy-slac) [inline provProto.py]: Why is UPDATE followed by INSERT? I expect one of them must fail.

File: drpPipe.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            objDecl = decl + random.uniform(-1, 1)
+            objFlux = random.uniform(0.01, 1)
+            # Insert object and source.
+            cursor.execute('''
+INSERT INTO Object(ra, decl, flux) VALUES (%f, %f, %f)''' % \
+                           (objRa, objDecl, objFlux))
+            objId = cursor.lastrowid
+            cursor.execute('''
+INSERT INTO Source(objectId, scExposureId, filter, ra, decl, flux)
+VALUES (%s, %s, "%s", %s, %s, %s)''' % \
```
Comment (andy-slac) [inline drpPipe.py:54]: We should prefer single quotes for string literals in SQL.

File: provProto.py (hunk 4 above)
Comment (jbecla) [inline provProto.py]: We need to set validityEnd for the config that was valid but it is validity ends, and we need to insert a new config. Why do you expect it to fail? We can and will have many configs for the same taskId.

File: drpPipe.py (hunk 5 above)
Comment (jbecla) [inline drpPipe.py:54]: will fix

File: calibPipe.py (hunk 1 above)
Comment (andy-slac) [inline calibPipe.py:2]: It's up to you of course, but it maybe harder to switch later than now.

File: calibPipe.py (hunk 2 above)
Comment (andy-slac) [inline calibPipe.py:63]: For very long literals it's probably worth doing that, for shorter once I'd just indent them further, e.g.:

```
    cursor.execute('''
            SELECT objectId, ra, decl, flux
            FROM   Object
            WHERE  ra BETWEEN %f AND %f
              AND  decl BETWEEN %f AND %f''' % (ra-1, ra+1, decl-1, decl+1))
```

File: provProto.py (hunk 4 above)
Comment (andy-slac) [inline provProto.py]: Sorry, I misread the schema. But this means that you are implementing conditions database for prv_cnf_Task, so you also need to know the time when you query this thing?

Comment (jbecla) [inline provProto.py]: "The time when I query this thing" is the time when a given exposure was processed ( == time when sources were detected on a given exposures).

Comment (andy-slac) [inline provProto.py]: Where does this time come from? Is it related to a particular TaskExecution?

Comment (jbecla) [inline provProto.py]: It is not related to task execution. Task execution has it's own datetime. prv_cnf_Task is keeping track of different configurations of prv_Task. So whatever things we are tracking per task go here. I added gitSHA as a placeholder for now. The idea is that when we have a new version of algorithm that runs as part of a given task, then it triggers a new config for that task, and that new config knows how to find that algorithm

File: calibPipe.py (hunk 2 above)
Comment (timj) [inline calibPipe.py:63]: Why do we substitute the variables into the string for all our SQL? Doesn't MySQL support placeholders that allow the SQL engine to do variable substitution with appropriate quoting internally?

Comment (jbecla) [inline calibPipe.py:63]: Right, I'll switch, good point. I keep forgetting about that every single time!

**Pull request merged**